apps/backend/src/app/api/admin/semesters/[id]/route.test.ts

```diff
@@ -74,6 +114,43 @@ describe("/api/admin/semesters/[id]", async () => {
       const json = await res.json()
       expect(json.error).toBe(getReasonPhrase(StatusCodes.INTERNAL_SERVER_ERROR))
     })
+
+    it("should rollback transaction if error occurs in transaction", async () => {
+      cookieStore.set(AUTH_COOKIE_NAME, adminToken)
+
+      const newSemester = await semesterDataService.createSemester(semesterCreateMock)
+      const newGameSessionSchedule = await gameSessionDataService.createGameSessionSchedule({
+        ...gameSessionScheduleMock,
+        semester: newSemester,
+      })
+      const newGameSession = await gameSessionDataService.createGameSession({
+        ...gameSessionMock,
+        gameSessionSchedule: newGameSessionSchedule,
+      })
+      const newBooking = await bookingDataService.createBooking({
+        ...bookingCreateMock,
+        gameSession: newGameSession,
+      })
+
+      vi.spyOn(SemesterDataService.prototype, "deleteRelatedDocsForSemester").mockRejectedValueOnce(
+        new Error("Cascade deletion failed"),
+      )
+
+      await DELETE({} as NextRequest, {
+        params: Promise.resolve({ id: "test" }),
+      })
+
+      const existingSemester = await semesterDataService.getSemesterById(newSemester.id)
+      const existingGameSessionSchedule = await gameSessionDataService.getGameSessionScheduleById(
+        newGameSessionSchedule.id,
+      )
+      const existingGameSession = await gameSessionDataService.getGameSessionById(newGameSession.id)
+      const existingBooking = await bookingDataService.getBookingById(newBooking.id)
+      expect(existingSemester).toBeDefined()
+      expect(existingGameSessionSchedule).toBeDefined()
+      expect(existingGameSession).toBeDefined()
+      expect(existingBooking).toBeDefined()
```

Comment:
i think we normally directly assert the value being defined in this case

```suggestion
      expect(await semesterDataService.getSemesterById(newSemester.id)).toBeDefined()
      expect(await gameSessionDataService.getGameSessionScheduleById(
        newGameSessionSchedule.id,
      )).toBeDefined()
      expect(await gameSessionDataService.getGameSessionById(newGameSession.id)).toBeDefined()
      expect(await bookingDataService.getBookingById(newBooking.id)).toBeDefined()
```